man/adoc/bpftrace.adoc

```diff
@@ -2783,3 +3524,247 @@ tracepoint:syscalls:sys_enter_openat
 uprobe:/bin/bash:rl_set_prompt
     const char *prompt
 ----
+
+=== Preprocessor Options
+
+The `-I` option can be used to add directories to the list of directories that bpftrace uses to look for headers.
+Can be defined multiple times.
+
+----
+# cat program.bt
+#include <foo.h>
+
+BEGIN { @ = FOO }
+
+# bpftrace program.bt
+
+definitions.h:1:10: fatal error: 'foo.h' file not found
+
+# /tmp/include
+foo.h
+
+# bpftrace -I /tmp/include program.bt
+
+Attaching 1 probe...
+----
+
+The `--include` option can be used to include headers by default.
+Can be defined multiple times.
+Headers are included in the order they are defined, and they are included before any other include in the program being executed.
+
+----
+# bpftrace --include linux/path.h --include linux/dcache.h \
+    -e 'kprobe:vfs_open { printf("open path: %s\n", str(((struct path *)arg0)->dentry->d_name.name)); }'
+
+Attaching 1 probe...
+open path: .com.google.Chrome.ASsbu2
+open path: .com.google.Chrome.gimc10
+open path: .com.google.Chrome.R1234s
+----
+
+=== Verbose Output
+
+The `-v` option prints more information about the program as it is run:
+
+----
+# bpftrace -v -e 'tracepoint:syscalls:sys_enter_nanosleep { printf("%s is sleeping.\n", comm); }'
+Attaching 1 probe...
+
+The verifier log:
+0: (bf) r6 = r1
+1: (b7) r1 = 0
+2: (7b) *(u64 *)(r10 -24) = r1
+3: (7b) *(u64 *)(r10 -32) = r1
+4: (7b) *(u64 *)(r10 -40) = r1
+5: (7b) *(u64 *)(r10 -8) = r1
+6: (7b) *(u64 *)(r10 -16) = r1
+7: (bf) r1 = r10
+8: (07) r1 += -16
+9: (b7) r2 = 16
+10: (85) call bpf_get_current_comm#16
+11: (79) r1 = *(u64 *)(r10 -16)
+12: (7b) *(u64 *)(r10 -32) = r1
+13: (79) r1 = *(u64 *)(r10 -8)
+14: (7b) *(u64 *)(r10 -24) = r1
+15: (18) r7 = 0xffff9044e65f1000
+17: (85) call bpf_get_smp_processor_id#8
+18: (bf) r4 = r10
+19: (07) r4 += -40
+20: (bf) r1 = r6
+21: (bf) r2 = r7
+22: (bf) r3 = r0
+23: (b7) r5 = 24
+24: (85) call bpf_perf_event_output#25
+25: (b7) r0 = 0
+26: (95) exit
+processed 26 insns (limit 131072), stack depth 40
+
+Attaching tracepoint:syscalls:sys_enter_nanosleep
+iscsid is sleeping.
+iscsid is sleeping.
+[...]
+----
+
+This includes `The verifier log:` and then the log message from the in-kernel verifier.
+
+== Advanced Topics
+
+=== Address Spaces
+
+Kernel and user pointers live in different address spaces which, depending on the CPU architecture, might overlap.
+Trying to read a pointer that is in the wrong address space results in a runtime error.
+This error is hidden by default but can be enabled with the `-kk` flag:
+
+----
+stdin:1:9-12: WARNING: Failed to probe_read_user: Bad address (-14)
+BEGIN { @=*uptr(kaddr("do_poweroff")) }
+        ~~~
+----
+
+bpftrace tries to automatically set the correct address space for a pointer based on the probe type, but might fail in cases where it is unclear.
+The address space can be changed with the <<functions-kptr, kptrs>> and <<functios-uptr, uptr>> functions.
+
+=== BTF Support
```

Comment:
Do you have some suggestions on the language here? I don't think I know enough about the different edge cases.

Reply:
How about something like:

```
It is not recommended to mix definitions from multiple sources (ie. BTF and header files). If your program mixes definitions, bpftrace will do its best but can easily get confused, as it has to go to great lengths to avoid redefinition conflicts. This logic may never be perfected.

Prefer to exclusively use BTF as it can never get out of sync on a running system. Almost all linux deployments will support BTF.
```

Probably better not to describe specific details. I forget a lot of the edge cases but I know we (mostly @viktormalik ) put quite a bit of effort into fixing it but could never get it perfect. I don't think it's worth the effort anyways -- BTF is far superior and almost always available.

Reply:
Thanks! Made a few light edits but sounds good to me.